opentelemetry-api/src/opentelemetry/context/base_context.py

```diff
@@ -67,64 +147,98 @@ def register_slot(
                 cls._slots[name] = cls.Slot(name, default)
             return cls._slots[name]
 
-    def apply(self, snapshot: typing.Dict[str, "object"]) -> None:
-        """Set the current context from a given snapshot dictionary"""
+    def merge(self, context: Context) -> Context:
+        new_context = self.current()
+        for key in context.snapshot:
+            new_context.snapshot[key] = context.snapshot[key]
+        return new_context
 
-        for name in snapshot:
-            setattr(self, name, snapshot[name])
-
-    def snapshot(self) -> typing.Dict[str, "object"]:
+    def _freeze(self) -> typing.Dict[str, "object"]:
         """Return a dictionary of current slots by reference."""
 
         keys = self._slots.keys()
         return dict((n, self._slots[n].get()) for n in keys)
 
     def __repr__(self) -> str:
-        return "{}({})".format(type(self).__name__, self.snapshot())
-
-    def __getattr__(self, name: str) -> "object":
-        if name not in self._slots:
-            self.register_slot(name, None)
-        slot = self._slots[name]
-        return slot.get()
-
-    def __setattr__(self, name: str, value: "object") -> None:
-        if name not in self._slots:
-            self.register_slot(name, None)
-        slot = self._slots[name]
-        slot.set(value)
+        return "{}({})".format(type(self).__name__, self._freeze())
 
-    def __getitem__(self, name: str) -> "object":
-        return self.__getattr__(name)
+    def _set(self, key: str, value: "object") -> None:
+        """
+        Set creates a Slot for the value if none exists, and sets the
+        value for that slot.
 
-    def __setitem__(self, name: str, value: "object") -> None:
-        self.__setattr__(name, value)
+        Args:
+            key: Name of the value
+            value: Contents of the value
+        """
+        if key not in self._slots:
+            self.register_slot(key, None)
+        slot = self._slots[key]
+        slot.set(value)
 
-    @contextmanager  # type: ignore
-    def use(self, **kwargs: typing.Dict[str, object]) -> typing.Iterator[None]:
-        snapshot = {key: self[key] for key in kwargs}
-        for key in kwargs:
-            self[key] = kwargs[key]
-        yield
-        for key in kwargs:
-            self[key] = snapshot[key]
+    def value(
+        self, key: str, context: typing.Optional["Context"] = None
+    ) -> typing.Optional["object"]:
```

Comment:
I find this method rather confusing. It's part of class `BaseContext` that implements `Context`, it receives a `context` parameter of type `Context` and, if `context` is passed the value is got from that, otherwise from self.

In other words, `A.do_work(par=B)` is actually doing work on `B` and not in `A`.